Propagate coordinator timeouts for regular writes and batches without throwing

piodul opened the pull request:
Currently, most of the failures that occur during CQL reads or writes are reported using C++ exceptions. Although the seastar framework avoids most of the cost of unwinding by keeping exceptions in futures as `std::exception_ptr`s, the exceptions need to be inspected at various points for the purposes of accounting metrics or converting them to a CQL error response. Analyzing the value and type of an exception held by `std::exception_ptr`'s cannot be done without rethrowing the exception, and that can be very costly even if the exception is immediately caught. Because of that, exceptions are not a good fit for reporting failures which happen frequently during overload, especially if the CPU is the bottleneck.

This PR introduces facilities for reporting exceptions as values using the boost::outcome library. As a first step, the need to use exceptions for reporting timeouts was eliminated for regular and batch writes, and no exceptions are thrown between creation of a `mutation_write_timeout_exception` and its serialization as a CQL response in the `cql_server`.

The types and helpers introduced here can be reused in order to migrate more exceptions and exception paths in a similar fashion.

Results of `perf_simple_query --smp 1 --operations-per-shard 1000000`:

    Master (00a9326ae79a8317055d7e5ad10b1ddff3613cc8)
    128789.53 tps ( 82.2 allocs/op,  12.2 tasks/op,   49245 insns/op)

    This PR
    127072.93 tps ( 82.2 allocs/op,  12.2 tasks/op,   49356 insns/op)

The new version seems to be slower by about 100 insns/op, fortunately not by much (about 0.2%).

Tests: unit(dev), unit(result_utils_test, debug)

piodul commented:
To measure the improvement, I disabled replica logic for two hardcoded tables by applying this diff:

```diff
diff --git a/service/storage_proxy.cc b/service/storage_proxy.cc
index 6fd630b5c..c46d7c6fe 100644
--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -2821,7 +2821,15 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
                 ++stats.read_repair_write_attempts.get_ep_stat(coordinator);
             }
 
-            if (coordinator == my_address) {
+            if (handler.get_schema()->ks_name() == "ks" && handler.get_schema()->cf_name() == "tbl_noop") {
+                // For ks.tbl_noop, the replica logic is always a noop
+                got_response(response_id, my_address, get_view_update_backlog());
+                f = make_ready_future<>();
+            } else if (handler.get_schema()->ks_name() == "ks" && handler.get_schema()->cf_name() == "tbl_fail") {
+                // For ks.tbl_fail, the replica logic immediately fails with a timeout
+                // Avoid rethrowing exceptions by inlining the latter exception handler logic here
+                got_failure_response(response_id, coordinator, forward_size + 1, std::nullopt, error::TIMEOUT, std::nullopt);
+            } else if (coordinator == my_address) {
                 f = futurize_invoke(lmutate);
             } else {
                 f = futurize_invoke(rmutate, coordinator, std::move(forward));
```

For `tbl_noop`, write operations always succeed without doing anything, and for `tbl_fail` they immediately fail due to a simulated timeout.

I ran a 10k write op/s workload for both master version (00a9326ae79a8317055d7e5ad10b1ddff3613cc8) and this PR. I ran the workload for the `tbl_noop` table first, then for `tbl_fail`. Here are some relevant grafana metrics for the __master__ version:

![with-exceptions](https://user-images.githubusercontent.com/4950936/152093716-fafb77bc-544d-402b-bc99-8bd83811e247.png)

And here are metrics for the version with the PR applied:

![without-exceptions](https://user-images.githubusercontent.com/4950936/152093733-2109dfd7-47b9-4ba7-a639-6ec3fbd9db91.png)

The `Load` metric changes as follows:

- Master version
	- `tbl_noop`: 21%
	- `tbl_fail`: 46%
- This PR
	- `tbl_noop`: 21%
	- `tbl_fail`: 24%

The impact on the happy path is not measurable (not graphanoleptically, at least). While there is some overhead associated with handling exceptions-as-values, it's much lower than using the old method.

dorlaor commented:
Wow, very impressive and important!

On Wed, Feb 2, 2022 at 6:51 AM Piotr Dulikowski ***@***.***>
wrote:

> To measure the improvement, I disabled replica logic for two hardcoded
> tables by applying this diff:
>
> diff --git a/service/storage_proxy.cc b/service/storage_proxy.cc
> index 6fd630b5c..c46d7c6fe 100644--- a/service/storage_proxy.cc+++ b/service/storage_proxy.cc@@ -2821,7 +2821,15 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
>                  ++stats.read_repair_write_attempts.get_ep_stat(coordinator);
>              }
>  -            if (coordinator == my_address) {+            if (handler.get_schema()->ks_name() == "ks" && handler.get_schema()->cf_name() == "tbl_noop") {+                // For ks.tbl_noop, the replica logic is always a noop+                got_response(response_id, my_address, get_view_update_backlog());+                f = make_ready_future<>();+            } else if (handler.get_schema()->ks_name() == "ks" && handler.get_schema()->cf_name() == "tbl_fail") {+                // For ks.tbl_fail, the replica logic immediately fails with a timeout+                // Avoid rethrowing exceptions by inlining the latter exception handler logic here+                got_failure_response(response_id, coordinator, forward_size + 1, std::nullopt, error::TIMEOUT, std::nullopt);+            } else if (coordinator == my_address) {
>                  f = futurize_invoke(lmutate);
>              } else {
>                  f = futurize_invoke(rmutate, coordinator, std::move(forward));
>
> For tbl_noop, write operations always succeed without doing anything, and
> for tbl_fail they immediately fail due to a simulated timeout.
>
> I ran a 10k write op/s workload for both master version (00a9326
> <https://github.com/scylladb/scylla/commit/00a9326ae79a8317055d7e5ad10b1ddff3613cc8>)
> and this PR. I ran the workload for the tbl_noop table first, then for
> tbl_fail. Here are some relevant grafana metrics for the *master* version:
>
> [image: with-exceptions]
> <https://user-images.githubusercontent.com/4950936/152093716-fafb77bc-544d-402b-bc99-8bd83811e247.png>
>
> And here are metrics for the version with the PR applied:
>
> [image: without-exceptions]
> <https://user-images.githubusercontent.com/4950936/152093733-2109dfd7-47b9-4ba7-a639-6ec3fbd9db91.png>
>
> The Load metric changes as follows:
>
>    - Master version
>       - tbl_noop: 21%
>       - tbl_fail: 46%
>    - This PR
>       - tbl_noop: 21%
>       - tbl_fail: 24%
>
> The impact on the happy path is not measurable (not graphanoleptically, at
> least). While there is some overhead associated with handling
> exceptions-as-values, it's much lower than using the old method.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/10014#issuecomment-1027575045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHURKLWD5K4EJTDAA7EC3UZCZ4RANCNFSM5NLA5OOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

gleb-cloudius commented on utils/exception_container.hh:
```diff
@@ -0,0 +1,136 @@
+            : _eptr(seastar::make_foreign(std::make_unique<exception_variant>(std::move(ex))))
+    { }
+
+    inline bool is_empty() const {
```
Why not `operator bool`?

piodul replied:
No good reason. I'll change it to `operator bool`.

elcallio replied:
STL conformance would be `bool empty() const` + operator.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -411,15 +411,15 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
         } else {
             if (_error == error::TIMEOUT) {
-                _ready.set_exception(mutation_write_timeout_exception(get_schema()->ks_name(), get_schema()->cf_name(), _cl, _cl_acks, _total_block_for, _type));
+                _ready.set_value(coordinator_failure(mutation_write_timeout_exception(get_schema()->ks_name(), get_schema()->cf_name(), _cl, _cl_acks, _total_block_for, _type)));
```
Do we need to go through the `coordinator_failure`. Is there c++ magic that can convert anything that inherits from std::exception into an error result?

piodul replied:
Oh, it turns out there are implicit constructors from the value type and error type. `_ready.set_value(mutation_write_timeout_exception(...))` just works. I'll remove the `coordinator_failure` function altogether.

I previously thought that `bo::failure()` was mandatory for constructing a result and tried doing `_ready.set_value(bo::failure(mutation_write_timeout_exception(...)))`, which didn't work so I wrote the helper function.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2137,7 +2137,7 @@ future<> storage_proxy::mutate_begin(unique_response_handler_vector ids, db::con
         send_to_live_endpoints(protected_response.release(), timeout); // response is now running and it will either complete or timeout
         return f;
-    });
+    }, result<>(bo::success()), utils::result_reducer{});
```
May be instead of `utils::result_reducer` you will provide `outcome_parallel_for_each` which hides the map_reduce details inside?

piodul replied:
Done. I named it `result_parallel_for_each` - the boost outcome library provides both `result` and `outcome` types, so I think a `result_` prefix generates less confusion.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2155,36 +2156,45 @@ future<result<>> storage_proxy::mutate_begin(unique_response_handler_vector ids,
     try {
-        mutate_result.get().value(); // .value() throws the error, if there is any
+        auto res = mutate_result.get();
+        if (res.has_error()) {
```
Can you just write `!res` here?

piodul replied:
Yes, it appears that the result type has a bool operator which works in this way. I'll change invocations of `has_error` to usages of the bool operator.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2517,18 +2517,33 @@ storage_proxy::mutate_atomically(std::vector<mutation> mutations, db::consistenc
+                };
+                if (f.failed()) {
+                    print_exception(f.get_exception());
+                } else if (result<> res = f.get(); res.has_error()) {
```
it should be enough to write `if (result<> res = f.get())` no?

piodul replied:
Not exactly, the boolean conversion returns true if the result holds a value and false in case of an error. We want to enter the branch if there is an error. I can change it to `if (result<> res = f.get(); !res)`.

gleb-cloudius replied:
Ah, right.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2517,18 +2517,33 @@ storage_proxy::mutate_atomically(std::vector<mutation> mutations, db::consistenc
+                if (f.failed()) {
+                    print_exception(f.get_exception());
+                } else if (result<> res = f.get(); res.has_error()) {
+                    res.assume_error().accept(print_exception);
```
Can we add `operator<<` to `coordinator_error`?

piodul replied:
Yes, sounds like a good idea. I'll use it here and in the `operator<<` for `result_message::exception` (it already has the logic for printing the exception).

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -2517,18 +2517,33 @@ storage_proxy::mutate_atomically(std::vector<mutation> mutations, db::consistenc
             return _p.mutate_prepare(_mutations, _cl, db::write_type::BATCH, _trace_state, _permit).then([this] (unique_response_handler_vector ids) {
-                return sync_write_to_batchlog().then([this, ids = std::move(ids)] () mutable {
+                return sync_write_to_batchlog().then([this, ids = std::move(ids)] (result<> res) mutable {
+                    if (res.has_error()) {
```
Yeah, all this ifs is not a nice part of using the `outcome` :(

piodul replied:
The `outcome` library defines macros which help reduce this kind of boilerplate ([link](https://www.boost.org/doc/libs/develop/libs/outcome/doc/html/tutorial/essential/result/try.html) to the relevant tutorial section). Maybe we could use them, and also define our versions which work inside functions that return futures? I don't usually like using macros, but perhaps the reduction in boilerplace is worth it.

There are two flavors, the first one is ugly but the second one requires a GCC/Clang extension (which, I guess is OK because we do not use other compilers):

```c++
BOOST_OUTCOME_TRY (auto i, BigInt::fromString(text));
// vs.
int i = BOOST_OUTCOME_TRYX (BigInt::fromString(text)); // It's an expression, so more flexible
```

gleb-cloudius replied:
Well, we do not like macros here, but I think meta progarmming is strong enough to provide something like:

```
template <typename L>
auto wrap(L lambda) {
   using ret_type = return_type<L>::value;
   using param_type = param_type<L>::value;

   return [lambda] (param_type v) -> ret_type {
        if (!v) {
            return v;
        }
        return lambda(v);
   }; 
}
```

avikivity replied:
I agree about disliking macros, I prefer a little extra boilerplate to them. Also agree a little metaprogramming (well-protected with concepts) can help.

piodul replied:
Done, added `utils::result_wrap`.

gleb-cloudius commented:
Overall less cumbersome that I would have imagined, but this only covers "simple" part of error propagation from a coordinator to a cql server. What about exceptions through below the coordinator level?

haaawk commented:
> Overall less cumbersome that I would have imagined, but this only covers "simple" part of error propagation from a coordinator to a cql server. What about exceptions through below the coordinator level?

We want to take one step at a time @gleb-cloudius. We may think about doing the same optimisation on the replica but that's out of the scope of this PR.

Also, this PR is just a spin-off from work on rate limiting that showed the performance problem of exceptions usage. The work in this PR will be mostly reused for rate limiting as well. The replica side is not that useful for rate limiting though so we want to push it for later and focus on rate limiting now as it is more important because users/customers are waiting for it.

piodul commented on service/storage_proxy.cc:
```diff
@@ -102,6 +102,19 @@ bo::failure_type<exceptions::coordinator_exception_container> coordinator_failur
 }
 
+template<typename T = void>
+future<result<T>> then_ok_result(future<T>&& f) {
```
I agree. I will move it there.

avikivity commented on test/boost/exception_container_test.cc:
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
```
Happy new year!

avikivity commented on utils/exception_container.hh:
```diff
@@ -0,0 +1,136 @@
+    // Bonus points: if each error type has a unique, globally-assigned
+    // identified integer, then conversion of the exception_container
+    // to a container supporting a superset of errors becomes very cheap.
+    seastar::foreign_ptr<std::unique_ptr<std::variant<Exs...>>> _eptr;
```
We can optimize slightly by using lw_shared_ptr and manufacturing the value ahead of time, since all the values will be duplicates of each other. I don't propose to do it now, since it's only a minor improvement and needs some care at the call sites.

avikivity commented on utils/exception_container.hh:
```diff
@@ -0,0 +1,136 @@
+    exception_container(const exception_container&) = delete;
+
+    exception_container& operator=(exception_container&&) = default;
+    exception_container& operator=(const exception_container&) = delete;
```
The move/copy constructors can be dropped, the compiler will synthesize the move constructors and delete the copy constructors itself.

piodul replied:
Done

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1430,7 +1430,7 @@ db::view::update_backlog storage_proxy::get_backlog_of(gms::inet_address ep) con
     auto& handler = _response_handlers.find(id)->second;
     handler->expire_at(timeout);
-    return handler->wait();
+    return handler->wait().then(utils::result_into_future<result<>>);
```
I guess response_wait will be converted to return future<result<>> later?

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1427,10 +1427,10 @@ db::view::update_backlog storage_proxy::get_backlog_of(gms::inet_address ep) con
 }
 
-future<> storage_proxy::response_wait(storage_proxy::response_id_type id, clock_type::time_point timeout) {
+future<result<>> storage_proxy::response_wait(storage_proxy::response_id_type id, clock_type::time_point timeout) {
```
Yes

avikivity commented on test/boost/cql_functions_test.cc:
```diff
@@ -66,6 +66,7 @@ SEASTAR_TEST_CASE(test_functions) {
                     }
                 }
                 virtual void visit(const result_message::bounce_to_shard& rows) override { throw "bad"; }
+                virtual void visit(const result_message::exception&) override { throw "bad"; }
```
Changelog calls it result_message::error, needs adjusting.

piodul replied:
Done

avikivity commented on transport/server.cc:
```diff
@@ -1156,6 +1170,8 @@ process_batch_internal(service::client_state& client_state, distributed<cql3::qu
             .then([stream, batch, q_state = std::move(q_state), trace_state = query_state.get_trace_state(), version] (auto msg) {
         if (msg->move_to_shard()) {
             return process_fn_return_type(dynamic_pointer_cast<messages::result_message::bounce_to_shard>(msg));
+        } else if (msg->is_exception()) {
```
Let's add __builtin_expect() to exception_container::is_empty() so that the compiler knows that this is an unexpected path and can optimize a little better.

piodul replied:
Done, added `__builtin_expect()` to the `empty` method (it was renamed from `is_empty` to `empty`).

avikivity commented:
Excellent series, very easy to review.

Please test with perf_simple_query --smp 1 --operations-per-shard 1000000, and report before/after. allocs/op, tasks/op, and insns/op are more stable measurements than ops/sec (though less meaningful) since they are less affected by CPU frequency scaling. You can add the test results to the cover letter.

nyh commented on transport/messages/result_message.cc:
```diff
@@ -22,6 +23,14 @@ std::ostream& operator<<(std::ostream& os, const result_message::bounce_to_shard
+    auto [name, what] = msg.get_exception().accept([] (const auto& ex) -> std::pair<const char*, const char*> {
+        return { typeid(ex).name(), ex.what() };
+    });
+    fmt::print(os, "{{result_message::exception {} ({})}}", what);
```
This operator<< probably shouldn't be in "result_message.cc" and should be in some general <util/outcome.hh> I suggested. Similar to how we have in Seastar's <util/log.hh>,
```
namespace std {
std::ostream& operator<<(std::ostream&, const std::exception_ptr&);
std::ostream& operator<<(std::ostream&, const std::exception&);
}
```
In your current code, maybe you can even use those existing operator<< instead of open-coding your own exception printing code.
By the way, one of the problems in your code here is that you print `typeid(ex).name()` without demangling the ugly C++-mangled names. At the minimum you should be using seastar::pretty_type_name() - but I think you can be using the entire operator<< from Seastar.

piodul replied:
That's a good point, I'll use the `operator<<` for the std::exception.

piodul replied:
Done

nyh commented on cql3/statements/batch_statement.cc:
```diff
@@ -24,6 +24,11 @@
 #include <boost/algorithm/cxx11/all_of.hpp>
 #include <boost/range/adaptor/uniqued.hpp>
 
+namespace bo = BOOST_OUTCOME_V2_NAMESPACE;
```
The more I see this "bo" thing the less I like it. As I noted above, I think even if we decided to use boost::outcome we should consider hiding it in our own <util/outcome.hh> which will allow us to avoid this weird boilerplate definitions.

nyh replied:
By the way, as far as I can tell (?) you aren't even using this "bo" namespace below, so it's not even needed.

piodul replied:
I'm using the convention that we already use for boost::intrusive:

```c++
using bi = boost::intrusive;
```

I'll remove the `using bo = ...` from this file.

piodul replied:
Oh, I think I parsed your second comment but not the first one. I'll move the `namespace bo` alias into the "utils/result.hh" header.

elcallio replied:
Should we really have namespace aliases in headers? Seems translation unit local to me...

piodul replied:
@elcallio @nyh I'm a bit confused, because to me it looks like your suggestions oppose each other. @nyh, perhaps I misinterpreted your comments - could you clarify what you would like me to do here?

nyh replied:
I didn't like the fact that we need to repeat this namespace command (which is particularly ugly in this case, with an ugly-looking macro) in a gazallion place. I agree with @elcallio that maybe calling it with the extremely short name "bo" in a header file is a bit too much. Perhaps a better idea is to hide the fact we are using Boost completely, in our own names - call it util::result and so on, and call the Boost stuff but later we could call std:: stuff when it's standardized, or implement our own.

Anyway, my comment here was just a minor nitpick, prompted by the ugly macro BOOST_OUTCOME_V2_NAMESPACE. I won't insist that you change anything here. It's not important enough to insist here.

piodul replied:
@nyh Thanks, I understand now. Outside "utils/result.hh", I think we mainly need the `bo` namespace for bo::success and bo::failure, so we could hide then as utils::result_success and utils::result_failure, respectively. For the time being, I would like to prioritize other work related to the "result migration", but after I'll no longer have conflicting work it will be a good idea to hide the `bo` namespace.

nyh commented on cql3/statements/modification_statement.cc:
```diff
@@ -259,23 +264,26 @@ modification_statement::do_execute(query_processor& qp, service::query_state& qs
     return get_mutations(qp, options, timeout, false, options.get_timestamp(qs), qs).then([this, cl, timeout, &qp, &qs] (auto mutations) {
         if (mutations.empty()) {
-            return now();
+            return make_ready_future<coordinator_result<>>(bo::success());
```
Again, is this bo::success() really needed? If it's not (nothing, or {}, can be used instead), you won't need to define this ugly namespace "bo" above.

piodul replied:
As explained in another discussion, `result<>` has its own default constructor explicitly deleted. For results holding `void` using `success()` seems to be the easiest way to initialize a result with value.

nyh commented on transport/server.cc:
```diff
@@ -866,16 +871,20 @@ cql_server::connection::process_on_shard(::shared_ptr<messages::result_message::
     ::shared_ptr<messages::result_message::bounce_to_shard>>;
 
+static inline coordinator_result<foreign_ptr<std::unique_ptr<cql_server::response>>> convert_error_message_to_coordinator_result(messages::result_message* msg) {
+    return bo::failure(std::move(*dynamic_cast<messages::result_message::exception*>(msg)).get_exception());
```
Can't we create a coordinator_result<> directly from the exception without using bo::failure?
If we could we could avoid using that ugly namespace in this source file.

By the way, I see you ended up using dynamic_cast<> as I suggested. So that's yet another reason not to use the std::variant<> and just use a common base class.

piodul replied:
The dynamic_cast here is not used to cast between exception types. It's used to cast a result_message into result_message::exception.

As for the `bo::failure`, you are right that we can create a coordinator_result<> directly from the exception. I'll remove `bo::failure`.

nyh commented on transport/server.cc:
```diff
@@ -431,9 +432,21 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
         });
         --_server._stats.requests_serving;
+
+        auto exception_handler = overloaded_functor {
```
I think (but maybe I'm wrong?) I asked in a previous patch what is this exception_handler, and it is only defined here. So I guess your patches need to be reordered.
Shouldn't this function be in some header file?

piodul commented:
Changelog v1 -> v2:

- `exception_container` now has `is_empty()` methods and `operator bool`,
- Got rid of `coordinator_failure` helper function
- Added `utils::result_parallel_for_each`
- Moved `then_ok_result` to `utils`
- Converted `res.has_value()` and `res.has_error()` expressions to use result's the `operator bool`.
- Added `operator<<` to `coordinator_exception`
- Added `utils::result_wrap`
- Removed move/copy constructors in `exception_container`
- Fixed the message of the commit which introduces `result_message::exception`
- Added `__builtin_expect` to `exception_container::empty()`
- Adjusted the `execute_xyz` family of methods in the `single_node_cql_env` so that they check for `result_message::exception` and throw it (this fixes auth_test.test_alter_with_timeouts)

The previous version can be viewed on the `pass-coordinator-timeouts-as-errors-v1` branch, can be seen [here](https://github.com/piodul/scylla/tree/pass-coordinator-timeouts-as-errors-v1).

piodul commented:
> Please test with perf_simple_query --smp 1 --operations-per-shard 1000000, and report before/after. allocs/op, tasks/op, and insns/op are more stable measurements than ops/sec (though less meaningful) since they are less affected by CPU frequency scaling. You can add the test results to the cover letter.

@avikivity I ran the benchmark and added results to the PR description. I included just one run, but locally I ran the test multiple times and the insns/op + tps metrics seem to fluctuate a bit. The number of allocations and tasks seem unchanged between versions and runs.

piodul commented on service/storage_proxy.cc:
```diff
@@ -88,8 +89,18 @@
+template<typename T = void>
+using result = service::storage_proxy::result<T>;
+
+template<typename Ex>
```
I removed `coodinator_result` and moved `then_ok_result` to the `utils` in v2, as per Gleb's suggestions. The latter was also moved to the commit which introduces the `utils/result.hh` file.

avikivity commented:
Great,  looks like we aren't paying a price for this.

piodul commented on service/storage_proxy.cc:
```diff
@@ -445,7 +445,7 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
         _proxy->_global_stats.background_write_bytes += _mutation_holder->size();
         _throttled = false;
-        _ready.set_value();
+        _ready.set_value(bo::success());
```
The default constructor for `result<>` is explicitly deleted. The author of the boost outcome library explains this decision in the FAQ: https://www.boost.org/doc/libs/1_71_0/libs/outcome/doc/html/faq.html#why-is-implicit-default-construction-disabled

nyh replied:
Sad. Yet another benefit to not using this library :-(
It sounds like the issue is that result<> also allows holding "nothing" - not a result and not an exception. I think this is a mistake (unless I missing something here). result<T> should hold T or an exception, not nothing. Moving a result<T> should be like moving T - and leave behind whatever moving T moves behind.

piodul replied:
> It sounds like the issue is that result<> also allows holding "nothing" - not a result and not an exception.

This was an issue with boost outcome v1's result<>. We are using v2, which does not have an empty state (from the FAQ: "v2 has eliminated any legal possibility of being empty"). I think this implies that the move constructor should work in the way you would like it to work.

nyh replied:
So I think these guys made a mistake in their judgement. Not only does it make sense for result<T>() to call T's default constructor, it's even more obvious this should be allowed when T is nothing (which I think is the case here) :-(
It's weird that we need to include that bizarre namespace just to spell out this "success()" instead of nothing.
But I guess it's not something we can fix now (if we use boost::outcome).

piodul commented on service/storage_proxy.cc:
```diff
@@ -2155,36 +2156,45 @@ future<result<>> storage_proxy::mutate_begin(unique_response_handler_vector ids,
-        stats.write_timeouts.mark();
-        return make_exception_future<>(std::current_exception());
+        exception_handler(ex);
+        return make_exception_future<result<>>(std::current_exception());
```
The new way of returning exceptions is meant to allow gradual migration. If somebody decides to migrate another exception type, they don't have to modify all throw sites at once - it is easier to focus on the most common cases first and leave others as throws. You also don't have to migrate the whole exception handling path at once, and you can add .then(result_into_future<...>) to convert a future<result<>> to future<> (which can generate an exceptional future).

It's just safer to cover both exceptions and results in the same place. Even if, after this PR, this function will never receive mutation_write_timeout_exception as a regular exception, it will protect us in the future if somebody starts throwing it or it appears due to some result -> exception future conversion logic. I wanted to set a precedent.

nyh replied:
Ok, while I understand this thinking, the downside is that this "precedent" means that exception-handling code now always (?) needs to be duplicated twice, because supposedly you never know if the exception will come from a result or an actual exception.

We had exactly the same problem with futures - a future-returning function can either return an exceptional future or really throw - and our solution was in places like then(), futurize(), etc., to catch exceptions and turn them into exceptional futures. Maybe here as well, the code could just catch the exception and turn it into a result, without any duplicated logic. Maybe this is what you were doing and I just misunderstood?

piodul replied:
> Ok, while I understand this thinking, the downside is that this "precedent" means that exception-handling code now always (?) needs to be duplicated twice, because supposedly you never know if the exception will come from a result or an actual exception.

Well, the pattern I used requires some code duplication, yes. Given exception type, you need to both handle it in result and write a catch() clause, so that's adding code to two places - and I can agree that it is fragile. The core of the exception handler is not duplicated as I moved it to the overloaded_functor and the catch() clause just calls it.

> We had exactly the same problem with futures - a future-returning function can either return an exceptional future or really throw - and our solution was in places like then(), futurize(), etc., to catch exceptions and turn them into exceptional futures. Maybe here as well, the code could just catch the exception and turn it into a result, without any duplicated logic. Maybe this is what you were doing and I just misunderstood?

I don't think it is possible to turn an exceptional future into a failed result<>. You can't, in general, take out an exception out of std::exception_ptr. If you rethrow it and do `catch (const std::exception& ex)`, the standard says that the reference is valid as long as the error handler is executing, so you can't return it from a function. On the other hand, you can't move out or copy the exception because it can be a derived class and you would have to somehow be able to dynamically invoke the correct copy constructor. Not to mention exceptions which are not derived from std::exception (we have catch (...) for that in the current logic).

I think I have an idea for a metaprogramming-powered way to handle results<>, exceptions and exceptional futures in a unified way. It would be used like this:

```c++
return utils::try_catch(
    [&] () { return some_operation_returning_future_result(); },
    result_catch<exception1>([&] (const auto& ex, auto&& rethrow) {
        return make_ready_future<result<>>(bo::success());
    }),
    result_catch<exception2>([&] (const auto& ex, auto&& rethrow) {
        // Handle the exception here
        // Because we don't want to convert between results and exceptions,
        // if we want to rethrow then we need to return different things.
        // Hence, the generic rethrow argument.
        return rethrow();
    }),
    result_catch_all([&] (auto&& rethrow) {
        // Equivalent to catch (...)
        return rethrow();
    })
);
```

I'm not sure if it won't be an overkill, though.

piodul replied:
I actually managed to write the `try_catch` function and it both looks and works nicely. However, it is definitely more complicated than the other result-related helper function and the change introducing it is not that small, so I would like to postpone it to a followup PR so that it can be reviewed independently.

nyh commented on utils/exception_container.hh:
```diff
@@ -0,0 +1,143 @@
+        os << typeid(ex).name() << " (" << ex.what() << ")";
+    });
+    return os;
+}
```
Copying my previous comment which got marked "outdated" by github:

I think you should use the existing operator<< from <seastar/utils/log.hh> instead of open-coding your own exception printing code.
One of the problems in your code here is that you print typeid(ex).name() without demangling the ugly C++-mangled names. At the minimum you should be using seastar::pretty_type_name() - but I think you can be using the entire operator<< from Seastar.

piodul replied:
Answered under the "outdated" comment, and done.

avikivity commented on exceptions/exceptions.hh:
```diff
@@ -19,9 +19,14 @@
 
 namespace exceptions {
 
+namespace bo = BOOST_OUTCOME_V2_NAMESPACE;
```
It's being [standardized](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0323r10.html) as std::expected. As soon as it hits, please post a patch to convert to std::expected.

Meanwhile, while I also dislike boost, it's better to use it than to spend time reimplementing it.

piodul commented:
Changelog v2 -> v2.1:

- Removed one use of `bo::success`
- Removed some `namespace bo` aliases from files which didn't need it
- exception_container's `operator<<` now uses std::exception's `operator<<`
- Adjusted the name of one test (`test_parallel_for_each` -> `test_result_parallel_for_each`)

The previous version can be viewed on the `pass-coordinator-timeouts-as-errors-v2` branch, [here](https://github.com/piodul/scylla/tree/pass-coordinator-timeouts-as-errors-v2).

piodul commented on utils/exception_container.hh:
```diff
@@ -0,0 +1,136 @@
+// Exceptions are held in a std::variant. The variant is kept behind
+// a std::unique_ptr to ensure that moves are cheap. This means that
+// a moved-out exception container becomes "empty" and does not contain
+// a valid exception.
```
I thought about your suggestion a bit more and I think it still is worth considering. While I like that exception_container explicitly lists possible exceptions, just keeping a shared_ptr\<std::exception\> or something in that fashion may be more consistent with how real C++ exceptions work.

That said, I'd rather try doing it in a followup. I don't think it will be too much work to convert the current approach to it, and the performance impact of the shared_ptr\<std::exception\> approach needs to be evaluated.

piodul commented:
Changelog v2.1 -> v2.2:

- Moved the `namespace bo = BOOST_OUTCOME_V2_NAMESPACE;` alias to the "utils/result.hh" header, removed from .cc files

The previous version can be viewed on the `pass-coordinator-timeouts-as-errors-v2.1` branch, [here](https://github.com/piodul/scylla/tree/pass-coordinator-timeouts-as-errors-v2.1).

avikivity commented:
I'll wait for @nyh to confirm his concerns were addressed. Meanwhile you can drop the RFC for the title, it is ready for merge.

avikivity commented:
btw, a note about benchmarking: it is safer to compare HEAD with $(git merge-base origin/master HEAD) rather than master, so you don't compare against unrelated changes in master. No need to redo it here, I don't think there was anything recent that touched performance, just a note for the future.

piodul commented:
> btw, a note about benchmarking: it is safer to compare HEAD with $(git merge-base origin/master HEAD) rather than master, so you don't compare against unrelated changes in master. No need to redo it here, I don't think there was anything recent that touched performance, just a note for the future.

No worries - in fact, what I called "master" here was the version off which I have based the PR. It was "master" when I created the PR's branch :) I'll be more explicit about it in the future.

```
[piodul@scyllapc scylla-4]$ git merge-base origin/master pass-coordinator-timeouts-as-errors
00a9326ae79a8317055d7e5ad10b1ddff3613cc8
```

piodul commented:
v2.2 -> v2.3: rebased on newest master.

Previous version is here: https://github.com/piodul/scylla/tree/pass-coordinator-timeouts-as-errors-v2.2

piodul commented:
Modified the title slightly to make it clear that the PR is about regular writes and batches only.

piodul commented:
> I'll wait for @nyh to confirm his concerns were addressed. Meanwhile you can drop the RFC for the title, it is ready for merge.

@nyh ping

nyh reviewed (approved):
@piodul you have addressed all my comments so no reason not to commit this.
If I was writing this code, then 1. I would have used inheritance (std::exception), not generics (std::variant) for the exception, and 2. I wouldn't have used boost - or at least hid its use in a compatibility layer. But I'm not writing this code ;-) And your approach is fine, and Avi seem to like it. So you  have my green light too.

piodul commented:
Please do not merge yet, I think there are still some places where `result_message::exception` might not get converted to an exception which can result in some exceptions being ignored. I need to double check and correct them.

piodul commented:
v3: reorganized the logic a bit to protect from `result_message::exception` being dropped accidentally

OK, so it turned out that my previous approach of propagating the exception through the CQL layer was naive. I added the `result_message::exception` and modified the code so that `cql_statement::execute` and `query_processor::execute_xyz` family of methods could return it in some cases. However, there are many other places in the code which call those functions and are not prepared at all to handle `result_message::exception` - which needs to be explicitly handled, otherwise the exception it carries is silently ignored. I thought all call sites used `result_message::visitor` to inspect it (which was modified to handle the `result_message::exception` case) but that was not the case.

I modified the code so that `cql_statement::execute` and `query_processor::execute_xyz` should behave as before, meanwhile I added `cql_statement::execute_without_checking_exception_message` and analogous methods for query_processor, which are allowed to return `result_message::exception` and their names clearly indicate that (at least I hope so). Unless I missed something, there shouldn't be a possibility of ignoring an exception.

I re-ran unit tests, re-checked that exceptions are indeed not thrown using the test in the cover letter. I re-ran the benchmark and results unfortunately got worse, but not by much (100 ins/op, ~0.2%).

Previous version can be found here: https://github.com/piodul/scylla/tree/pass-coordinator-timeouts-as-errors-v2.3

piodul commented:
@avikivity / @nyh / @gleb-cloudius could you take a look once more? I think I eliminated the possibility of result_message::exception being ignored, but I would appreciate if somebody else confirmed it.

Commits before "[result_message: add result_message::exception](https://github.com/scylladb/scylla/pull/10014/commits/6f4aae025d8fdb6874be750eacd5b607b5321603)" were untouched.

nyh commented on test/lib/cql_test_env.cc:
```diff
@@ -249,8 +253,10 @@ class single_node_cql_env : public cql_test_env {
-            .finally([qs, qo = std::move(qo)] {});
+        return local_qp().execute_prepared_without_checking_exception_message(std::move(prepared), std::move(id), *qs, lqo, true)
+            .then([qs, qo = std::move(qo)] (auto msg) {
+                return cql_transport::messages::propagate_exception_as_future(std::move(msg));
```
But this is just the testing environment... How is it relevant to Scylla performance? I guess you are running your test in the testing environment instead of real Scylla?

piodul replied:
I meant the performance of the `perf_simple_query` benchmark. Yes, this is just testing code and does not have impact on the real Scylla, so maybe the regression I measured is completely virtual (on the other hand maybe there is some undetected regression outside the code path tested by the benchmark...).

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -241,6 +244,12 @@ static thread_local inheriting_concrete_execution_stage<
 modification_statement::execute(query_processor& qp, service::query_state& qs, const query_options& options) const {
+    return execute_without_checking_exception_message(qp, qs, options)
+            .then(cql_transport::messages::propagate_exception_as_future<shared_ptr<cql_transport::messages::result_message>>);
+}
```
Is there an explanation why this did not add an allocation+task? The future returned by execute_without_checking_exception_message() could not be ready, and so this adds a continuation.

Perhaps we eliminated a continuation from some other place.

avikivity replied:
Oh wait, the regular path calls execute_without_checking_exception_message().

avikivity commented:
Dequeued, result_utils_test fails in debug mode.

piodul commented:
v3.1

Found the reason the test failed in debug mode. In `test_then_ok_result`, I'm using `utils::then_ok_result` which takes a future and calls `.then()`. I always passed a ready/exceptional future, so I thought a task would not get allocated and it was safe to call `.get()` on the future without running in a seastar thread, but it looks like this optimization is disabled in debug mode.

I converted all tests in `result_utils_test` which call `future::get()` so that they run in a seastar thread.

piodul commented:
The result_utils_test now passes in debug mode. @avikivity , is there anything else you would like me to do here or is it OK to merge again?

piodul commented:
@avikivity Thanks!

piodul commented:
It turns out that the benchmark I used to run ran **reads**, not writes. For clear conscience I re-ran the benchmark in write mode with pre-merge and post-merge Scylla and compared performance. The difference is not as good as described in the cover letter:

```
126962.70 tps ( 66.2 allocs/op,  13.2 tasks/op,   51359 insns/op)
128654.00 tps ( 66.2 allocs/op,  13.2 tasks/op,   51344 insns/op)
129517.18 tps ( 66.2 allocs/op,  13.2 tasks/op,   51348 insns/op)
128744.86 tps ( 66.2 allocs/op,  13.2 tasks/op,   51338 insns/op)
129800.16 tps ( 66.2 allocs/op,  13.2 tasks/op,   51338 insns/op)

127608.55 tps ( 67.2 allocs/op,  14.2 tasks/op,   52407 insns/op)
127985.63 tps ( 67.2 allocs/op,  14.2 tasks/op,   52414 insns/op)
127386.62 tps ( 67.2 allocs/op,  14.2 tasks/op,   52415 insns/op)
127882.66 tps ( 67.2 allocs/op,  14.2 tasks/op,   52407 insns/op)
127396.76 tps ( 67.2 allocs/op,  14.2 tasks/op,   52403 insns/op)
```

It seems that there is an additional allocation and a task now, and we use ~1050 instructions more.

/cc @avikivity

avikivity commented:
Let's identify and fix it.

piodul commented:
@avikivity I managed to recover the task/allocation and some insns by optimizing the `utils::result_parallel_for_each`: https://github.com/scylladb/scylla/pull/10053
